### PR TITLE
feat(spanner): add support for change streams transaction exclusion option

### PR DESCRIPTION
**Pull request by alkatrivedi**
Add support for excluding transactions from being recorded in the change streams by passing a new boolean option ExcludeTxnFromChangeStreams in the various write APIs:

`runTransaction`
`getTransaction`
`runPartitionedUpdate`
`_mutate`

Note: Samples will be added later in separate prs.

### CODE REVIEW

File: src/database.ts [hunk 1]
```diff
@@ -62,6 +62,7 @@ import {CreateTableCallback, CreateTableResponse, Table} from './table';
 import {
   ExecuteSqlRequest,
   RunCallback,
+  runPartitionedUpdateOptions,
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/database.ts]: Done

File: src/database.ts [hunk 2]
```diff
@@ -2732,33 +2738,44 @@ class Database extends common.GrpcServiceObject {
...
-      }
+    if (typeof query === 'string') {
+      query = {sql: query} as runPartitionedUpdateOptions;
+    }
+
+    query = Object.assign({}, query) as runPartitionedUpdateOptions;
 
-      transaction.runUpdate(query, (err, updateCount) => {
+    transaction.begin(
+      {
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/database.ts]: Done

File: src/transaction.ts [hunk 3]
```diff
@@ -388,11 +397,14 @@ export class Snapshot extends EventEmitter {
    *   });
    * ```
    */
-  begin(gaxOptions?: CallOptions): Promise<BeginResponse>;
+  begin(gaxOptions?: BeginTransactionOptions): Promise<BeginResponse>;
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/transaction.ts]: Yeah, thanks for pointing it out. I have made the changes

File: src/table.ts [hunk 4]
```diff
@@ -46,6 +46,7 @@ export type DropTableCallback = UpdateSchemaCallback;
 
 interface MutateRowsOptions extends CommitOptions {
   requestOptions?: Omit<IRequestOptions, 'requestTag'>;
+  excludeTxnFromChangeStreams?: true;
```
Comment (surbhigarg92) [inline src/table.ts:49]: Change this default value

File: src/database.ts [hunk 5]
```diff
@@ -62,6 +62,7 @@ import {CreateTableCallback, CreateTableResponse, Table} from './table';
 import {
   ExecuteSqlRequest,
   RunCallback,
+  RunPartitionedUpdateOptions,
```
Comment (surbhigarg92) [inline src/database.ts:65]: Move this option in this class

**Review by surbhigarg92 — APPROVED** [review]

File: src/database.ts (hunk 5 above)
Comment (alkatrivedi) [reply src/database.ts]: Done

**Review by olavloite — APPROVED** [review]

File: src/table.ts [hunk 6]
```diff
@@ -1079,6 +1080,15 @@ class Table {
...
       }
 
+      const excludeTxnFromChangeStreams =
+        'excludeTxnFromChangeStreams' in options
+          ? options.excludeTxnFromChangeStreams
+          : false;
+
+      if (excludeTxnFromChangeStreams) {
+        transaction!.excludeTxnFromChangeStreams();
+      }
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply src/table.ts:1090]: Yes, we need this `excludeTxnFromChangeStreams` in the `option` for the cases where a customer can make a simple mutation calls without using transaction object.
Example: insertion using table name `singersTable.insert()`

File: test/spanner.ts [hunk 7]
```diff
@@ -3074,6 +3074,34 @@ describe('Spanner with mock server', () => {
         await database.close();
       });
 
+      it('should retry on aborted error with excludeTxnFromChangeStreams', async () => {
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply test/spanner.ts:3077]: Yeah, we don't need this test case. I have removed it.

File: test/spanner.ts [hunk 8]
```diff
@@ -3074,6 +3074,34 @@ describe('Spanner with mock server', () => {
...
+            streamIndex: 1,
+          } as MockError)
+        );
+        const [updateCount] = await database.runPartitionedUpdate({
+          sql: updateSql,
+          excludeTxnFromChangeStreams: true,
+        });
+        assert.strictEqual(updateCount, 2);
+
+        const beginTxnRequest = spannerMock.getRequests().find(val => {
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply test/spanner.ts:3094]: We do not need this test either, hence removing it too.

File: test/spanner.ts [hunk 9]
```diff
@@ -3279,176 +3422,231 @@ describe('Spanner with mock server', () => {
...
-    it('should only inline one begin transaction', async () => {
-      const database = newTestDatabase();
-      await database.runTransactionAsync(async tx => {
-        const rowCount1 = getRowCountFromStreamingSql(tx!, {sql: selectSql});
-        const rowCount2 = getRowCountFromStreamingSql(tx!, {sql: selectSql});
-        await Promise.all([rowCount1, rowCount2]);
-        await tx.commit();
-      });
-      await database.close();
+  it('should reuse a session for optimistic and pessimistic locks', async () => {
```
[the remark this answers is not in the document]
Comment (alkatrivedi) [reply test/spanner.ts]: It was a formatting issue, I have fixed it.

**Pull request merged**